compiler/rustc_session/src/options.rs

```diff
@@ -58,10 +58,58 @@ macro_rules! hash_substruct {
     };
 }
 
+macro_rules! gather_tmods {
+    ($_opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [SUBSTRUCT], [TMOD]) => {
+        compile_error!("SUBSTRUCT can't be TMOD (target modifier)");
+    };
+    ($opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [UNTRACKED], [TMOD]) => {
+        if *$opt_expr != $init {
+            $mods.push(format!("{}={:?}", stringify!($opt_name), $opt_expr));
+        }
+    };
+    ($opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [TRACKED], [TMOD]) => {
+        if *$opt_expr != $init {
+            $mods.push(format!("{}={:?}", stringify!($opt_name), $opt_expr));
+        }
+    };
+    ($opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [TRACKED_NO_CRATE_HASH], [TMOD]) => {
+        if *$opt_expr != $init {
+            $mods.push(format!("{}={:?}", stringify!($opt_name), $opt_expr));
+        }
+    };
+    ($_opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [SUBSTRUCT], []) => {
+        $opt_expr.gather_target_modifiers($mods);
+    };
+    ($_opt_name:ident, $init:expr, $_opt_expr:expr, $_mods:expr, [UNTRACKED], []) => {{}};
+    ($_opt_name:ident, $init:expr, $_opt_expr:expr, $_mods:expr, [TRACKED], []) => {{}};
+    ($_opt_name:ident, $init:expr, $_opt_expr:expr, $_mods:expr, [TRACKED_NO_CRATE_HASH], []) => {{}};
+}
+
+macro_rules! gather_tmods_top_level {
+    ($_opt_name:ident, $opt_expr:expr, $mods:expr, [SUBSTRUCT], [TMOD]) => {
+        compile_error!("SUBSTRUCT can't be TMOD (target modifier)");
+    };
+    ($opt_name:ident, $opt_expr:expr, $mods:expr, [UNTRACKED], [TMOD]) => {
+        compile_error!("Top level option can't be TMOD (target modifier)");
```

Comment:
Why not?

Reply:
Well, top-level options are very specific (in code), they don't have default value, some top-level options are copied from cg/unstable. Anyway, currently it looks like abi/vulnerability options are in codegen or unstable.